pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -748,3 +752,152 @@ func (os *OpenStack) GetZone() (cloudprovider.Zone, error) {
+		Region: os.region,
+	})
+
+	if err != nil || cClient == nil {
+		glog.Errorf("Unable to initialize nova client for region: %s", os.region)
+		return "", err
+	}
+
+	compute_id, err := os.getComputeIDbyHostname(cClient)
+	if err != nil || len(compute_id) == 0 {
+		glog.Errorf("Unable to get minion's id by minion's hostname")
+		return "", err
+	}
+
+	if len(disk.Attachments) > 0 && disk.Attachments[0]["server_id"] != nil {
+		if compute_id == disk.Attachments[0]["server_id"] {
+			glog.Infof("Disk: %q is already attached to compute: %q\n", diskName, compute_id)
+			return disk.ID, nil
+		} else {
+			errMsg := fmt.Sprintf("Disk %q is attached to a different compute: %q, should be detached before proceeding", diskName, disk.Attachments[0]["server_id"])
+			glog.Errorf(errMsg)
+			return "", errors.New(errMsg)
+		}
+	}
+	// add read only flag here if possible spothanis
+	_, err = volumeattach.Create(cClient, compute_id, &volumeattach.CreateOpts{
+		VolumeID: disk.ID,
+	}).Extract()
+	if err != nil {
+		glog.Infof("Failed to attach %s volume to %s compute", diskName, compute_id)
+		return "", err
+	}
+	glog.Infof("Successfully attached %s volume to %s compute", diskName, compute_id)
+	return disk.ID, nil
+}
+
+func (os *OpenStack) DetachDisk(partialDiskId string) error {
+	disk, err := os.getVolume(partialDiskId)
+	if err != nil {
+		return err
+	}
+
+	cClient, err := openstack.NewComputeV2(os.provider, gophercloud.EndpointOpts{
+		Region: os.region,
+	})
+
+	compute_id, err := os.getComputeIDbyHostname(cClient)
+	if err != nil || len(compute_id) == 0 {
+		glog.Errorf("Unable to get minion's ip while detaching disk")
+		return err
+	}
+	if len(disk.Attachments) > 0 && disk.Attachments[0]["server_id"] != nil && compute_id == disk.Attachments[0]["server_id"] {
+		err = volumeattach.Delete(cClient, compute_id, disk.ID).ExtractErr()
+		if err != nil {
+			glog.Errorf("Failed to delete volume %s from compute %s attached %v\n", disk.ID, compute_id, err)
+			return err
+		}
+	} else {
+		errMsg := fmt.Sprintf("Disk: %s has no attachments or is not attached to compute: %s\n", disk.Name, compute_id)
+		glog.Errorf(errMsg)
+		return errors.New(errMsg)
+	}
+	return nil
+}
+
+//Takes a partial/full disk id or diskname
+
+func (os *OpenStack) getVolume(diskName string) (volumes.Volume, error) {
+	sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+		Region: os.region,
+	})
+
+	var volume volumes.Volume
+	if err != nil || sClient == nil {
+		glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+		return volume, err
+	}
+
+	err = volumes.List(sClient, nil).EachPage(func(page pagination.Page) (bool, error) {
+		vols, err := volumes.ExtractVolumes(page)
+		if err != nil {
+			glog.Errorf("Failed to extract volumes: %v\n", err)
+			return false, err
+		} else {
+			for _, v := range vols {
+				glog.Infof("%s\t%s\t%v\n", v.ID, v.Name, v.Attachments)
```

Comment:
Does this always need to be logged? Would it make sense to reduce verbosity to `V(4)` or `V(5)`?